deep_qa/models/sentence_selection/siamese_sentence_selector.py

```diff
@@ -0,0 +1,115 @@
+from typing import Any, Dict
+from overrides import overrides
+from keras.layers import Input
+
+from ...data.instances.sentence_selection_instance import SentenceSelectionInstance
+from ...layers.attention.attention import Attention
+from ...layers.wrappers.encoder_wrapper import EncoderWrapper
+from ...training.text_trainer import TextTrainer
+from ...training.models import DeepQaModel
+
+
+class SiameseSentenceSelector(TextTrainer):
+    def __init__(self, params: Dict[str, Any]):
+        self.max_question_length = params.pop('max_question_length', None)
+        self.max_sentence_length = params.pop('max_sentence_length', None)
+        self.num_sentences = params.pop('num_sentences', None)
+        super(SiameseSentenceSelector, self).__init__(params)
+
+    @overrides
+    def _build_model(self):
+        """
+        The basic outline here is that we'll pass the questions and each
+        sentence in the passage through some sort of encoder. At the simplest,
+        this could be a biLSTM or a biGRU.
+
+        Then, we take the encoded representation of the question and calculate
+        a cosine similarity with the encoded representation of each sentence in
+        the passage, to get a tensor of cosine similarities with shape
+        (batch_size, num_sentences_per_passage). We then normalize for each
+        batch to get a probability distribution over sentences in the passage.
+        """
+        # First we create input layers and pass the inputs through embedding layers.
+        # shape: (batch size, question_length)
+        question_input = Input(shape=self._get_sentence_shape(self.max_question_length),
+                               dtype='int32', name="question_input")
+
+        # shape: (batch size, num_sentences, sentences_length)
+        sentences_input_shape = ((self.num_sentences,) +
+                                 self._get_sentence_shape(self.max_sentence_length))
+        sentences_input = Input(shape=sentences_input_shape,
+                                dtype='int32', name="sentences_input")
+
+        # shape: (batch size, question_length, embedding size)
+        question_embedding = self._embed_input(question_input)
+
+        # shape: (batch size, num_sentences, sentences_length, embedding size)
+        sentences_embedding = self._embed_input(sentences_input)
+
+        # We encode the question embeddings with some encoder.
+        question_encoder = self._get_encoder(name="question_encoder")
```

Comment:
Add a `fallback_behavior` here and below, probably `use_default_encoder` (that way you can have them use the same encoder if you don't specify separate ones).